tensorflow/lite/kernels/internal/reference/integer_ops/exp.h

```diff
@@ -0,0 +1,93 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_LITE_KERNELS_INTERNAL_REFERENCE_INTEGER_OPS_EXP_H_
+#define TENSORFLOW_LITE_KERNELS_INTERNAL_REFERENCE_INTEGER_OPS_EXP_H_
+
+#include "public/gemmlowp.h"
+#include "tensorflow/lite/kernels/internal/common.h"
+#include "tensorflow/lite/kernels/internal/types.h"
+
+namespace tflite {
+namespace reference_integer_ops {
+
+template <typename T>
+inline void QExp(const ExpParams& params, const RuntimeShape& input_shape,
+                 const T* input_data, const RuntimeShape& output_shape,
+                 T* output_data) {
+  const int32 input_zero_point = params.input_zero_point;
+  const int32 input_range_radius = params.input_range_radius;
+  const int32 input_multiplier = params.input_multiplier;
+  const int input_left_shift = params.input_left_shift;
+  const int32 output_zero_point = params.output_zero_point;
+  const int size = MatchingFlatSize(input_shape, output_shape);
+
+  int c = 0;
+  for (; c < size; ++c) {
+    const T input_val = input_data[c];
+    const int32 input_val_centered =
+        static_cast<int32>(input_val) - input_zero_point;
+    // For all x inside the radius, evaluate exp on (-input_range_radius, 0).
+    T output_val;
+    if (input_val_centered < -input_range_radius) {
+      output_val = std::numeric_limits<T>::min();
+    } else if (input_val_centered > input_range_radius) {
+      output_val = std::numeric_limits<T>::max();
+    } else {
+      const int32 input_val_rescaled =
+          MultiplyByQuantizedMultiplierGreaterThanOne(
+              input_val_centered, input_multiplier, input_left_shift);
+
+      using FixedPoint4 = gemmlowp::FixedPoint<int32, 4>;
+      using FixedPoint0 = gemmlowp::FixedPoint<int32, 0>;
+      const FixedPoint4 input_val_f4 = FixedPoint4::FromRaw(input_val_rescaled);
+      FixedPoint0 output_val_f0;
+      using gemmlowp::RoundingDivideByPOT;
+      int32 output_val_s32;
+      if (input_val_centered > 0) {
+        // For x > 0, calculate exp(x) as 1/(exp(-x))
+        output_val_f0 = gemmlowp::exp_on_negative_values(-input_val_f4);
+        int num_bits_over_unit;
+        FixedPoint0 shifted_scale = FixedPoint0::FromRaw(
+            GetReciprocal(output_val_f0.raw(), 0, &num_bits_over_unit));
+        output_val_s32 = RoundingDivideByPOT(
```

Comment:
Thanks for refactoring. Now that the code is structurally in a good place, we can take a closer look at the numerics. Here I see what is essentially done is: Take exp(-x) and get the reciprocal. This value is right shifted (with rounding) by `input_left_shift + 3 + num_bits_over_unit` bits. Why the value `3` and not some other value? This needs more explanation. Additionally, if this shift amount is acceptable for uint8 and int8 quantization, there should be a comment for the quantization strategy for both (i.e. the fixed point ranges for both cases where this exp calculation is well defined, etc.).

Do you have a model that uses this op that could be a part of this submission? If not this op directly, maybe a model that uses quantized Elu? If you have such a model and we can run it with TfLite tools to see accuracy, it makes it more likely to accept this submission because we can make sure that future changes don't regress the accuracy. Additionally, these choices of number of bits for shifts are usually tied tightly to the actual quantization strategy, so it is difficult to accept an Op without the larger context of the quantization strategy forcing certain choices (like this `3` here).

Reply:
@talumbau , sorry for late response, I have updated the code with the logic of 27 this is because in Prepare function we have  choose integer bits to be 4
So based on that we need to right shift 27 plus number of bits we got from GetReciprocal API. Hope this clears the things

Also as mentioned this is required by issue #28438.
Regards
Amit